for_dict_challenges_bonus.py

```diff
@@ -65,6 +65,92 @@ def generate_chat_history():
+        if message['reply_for'] is not None:  # если ответ на сообщение пользователя не None
+            parent_message = message  # в переменную parent_message передаётся весь message
+            user_id = parent_message['sent_by']  # в переменную user_id передаётся ID пользователя
+            users_counter[user_id] = users_counter.get(user_id, 0) + 1  # создаётся словарь ключ (ID пользователя): значение (количество сообщений, направленных ему)
+    
+    user_id_max = max(users_counter.items(), key=lambda items: items[1])  # определяется кортеж (ID пользователя, максимальное количество сообщений)
+    
+    users_id_max = [] 
+    for key, value in users_counter.items():  # для определения всех айди пользователей (в случае повторений максимального количества сообщений), на сообщения которого больше всего отвечали
+        if value == user_id_max[1]:
+            users_id_max.append(key)
+
+    return f'ID пользователя(ей), на сообщения которого(ых) больше всего отвечали: {users_id_max}'
+
+# 3. Вывести айди пользователей, сообщения которых видело больше всего уникальных пользователей.
+def id_users_posts_seen_most_uniq_users(history):
+    messages_db = {}
+    for message in history:
+        messages_db[message['id']] = message
+    
+    id_user_uniq_users = {}  # создаём словарь ключ (id_messege): значение (количество уникальных пользователей)
+    for mess in messages_db.values():
+        id_user_uniq_users[mess['id']] = mess['seen_by'] = len(mess['seen_by']) # замена значения по ключу seen_by в словаре messages_db
+    
+    num_max_uniq_user = max(id_user_uniq_users.items(), key=lambda items: items[1])  # определяется кортеж (id_messege, максимальное количество уникальных пользователей)
+
+    # users_id_max = {}  можно создать словарь и подсчитать максимальное количество уникальных просмотров с учётом повторяющихся id_user
+    users_id_max = set()  # создаём множество, чтобы не дублировались одинаковые id_user
+    for message in messages_db.values():
+        if message['seen_by'] == num_max_uniq_user[1]:
+            # users_id_max[message['sent_by']] = users_id_max.get(message['sent_by'], 0) + 1
+            users_id_max.add(message['sent_by'])
+      
+    return f'ID пользователей, сообщения которых видело больше всего уникальных пользователей: {users_id_max}'
+
+# 4. Определить, когда в чате больше всего сообщений: утром (до 12 часов), днём (12-18 часов) или вечером (после 18 часов).
+
+def time_most_messages_in_chat(history):
+    # Период времени с 00:00:00 по 05:59:59 не учитывается в соответствии с заданием
+
+    num_all_messages = {                  # словарь - ключ (период времени): значение (id сообщений за это время). 
+            'утром (до 12 часов)': [],
+            'днём (12-18 часов)': [],
+            'вечером (после 18 часов)': []
+            }
+
+    for message in history:
+        if '06:00:00' <= message['sent_at'].strftime('%H:%M:%S') <= '12:00:00':  # если время сообщения больше заданного
```

Comment:
Тоже работает 😃, но можно проще сделать. Так как  sent_at это datetime, ты можешь обращаться к его атрибуту hours. 
```suggestion
        if 6 <= message['sent_at'].hours <= 11:  # если время сообщения больше заданного
```